src/main/java/jenkins/plugins/gerrit/client/GerritClientBuilder.java

```diff
@@ -0,0 +1,177 @@
+    }
+    else if (envVars.containsKey("GERRIT_CHANGE_URL")) {
+      gerritApiUrl(new GerritURI(new URIish(envVars.get("GERRIT_CHANGE_URL"))).getApiURI());
+    }
+    insecureHttps(Boolean.parseBoolean(envVars.get("GERRIT_API_INSECURE_HTTPS")));
+    String credentialsId = envVars.get("GERRIT_CREDENTIALS_ID");
+    if (credentialsId != null) {
+      StandardUsernamePasswordCredentials credentials = com.cloudbees.plugins.credentials.CredentialsProvider.findCredentialById(credentialsId,
+              StandardUsernamePasswordCredentials.class, context.get(Run.class));
+      credentialsProvider(new UsernamePasswordCredentialsProvider(credentials));
+    }
+    return this;
+  }
+
+  public GerritClient build() throws IOException {
+    if (gerritApiUrl == null) {
+      logger.println("Gerrit Review is disabled no API URL");
+      return null;
+    }
+    if (credentialsProvider == null) {
+      logger.println("Gerrit Review is disabled no credentials");
+      return null;
+    }
+    UsernamePasswordCredentialsProvider.UsernamePassword usernamePassword = credentialsProvider.getUsernamePassword(gerritApiUrl);
+    if (usernamePassword.username == null) {
+      logger.println("Gerrit Review is disabled no credentials");
+      return null;
+    }
+    boolean useRest;
+    switch (gerritApiUrl.getScheme()) {
+      case "http":
+      case "https":
+        useRest = true;
+        break;
+      case "ssh":
+        useRest = false;
+        break;
+      default:
+        throw new IllegalStateException("Unknown scheme " + gerritApiUrl.getScheme());
+    }
+    if (useRest) {
+      GerritAuthData.Basic authData =
+              new GerritAuthData.Basic(gerritApiUrl.toString(), usernamePassword.username, usernamePassword.password);
+      List<HttpClientBuilderExtension> extensions = new ArrayList<>();
+      if (Boolean.TRUE.equals(insecureHttps)) {
+        extensions.add(SSLNoVerifyCertificateManagerClientBuilderExtension.INSTANCE);
+      }
+      GerritRestApi gerritRestApi = new GerritRestApiFactory()
+              .create(authData, extensions.toArray(new HttpClientBuilderExtension[0]));
+      return new GerritClientRest(gerritRestApi, gerritApiUrl.getScheme());
+    } else {
+      RemoteSession remoteSession;
+      try {
+
+        //todo: JSch.setConfig //Strict Auth should be optional
+        //todo: Prettier logging
+        //todo: SSHCredentialsPlugin
+
+        java.util.logging.Logger sshLogger = java.util.logging.Logger.getLogger("SSH");
+        sshLogger.setLevel(Level.ALL);
+        JSch.setLogger(new Logger() {
+          @Override
+          public boolean isEnabled(int level) {
+            Level l = getLevel(level);
+            return sshLogger.isLoggable(l);
+          }
+
+          @Override
+          public void log(int level, String message) {
+            Level l = getLevel(level);
+            sshLogger.log(l, message);
+          }
+
+          private Level getLevel(int level) {
```

Comment:
map of Intgeger->Level will be much better, however, I do not know what is the best practice for logging in jenkins... I thought that printing into the listener logger is the way to go.

Reply:
Yeah I plan to redo the logging-mapping completely. It was just hastily implemented to get some output from the SSH lib.